### PR TITLE
ExecOpRequestSPARQLWithTranslation

### PR DESCRIPTION
**Pull request by JulietteVV**

### CODE REVIEW

**Comment by JulietteVV** [comment]
@scferrada About the original request, you mentioned that it could not be a union but could it be a SPARQLGroupPattern? Or was it only TriplePattern and BGP?

**Comment by scferrada** [comment]
I think it should only be TriplePattern or BGP, because we have defined the translation method only for those two.

However, it seems that translating unions (of TriplePatterns and BGPs) is straightforward, right? In which case we could leave that code, and maybe mention it in the thesis, if it is not too long.

**Comment by JulietteVV** [comment]
> I think it should only be TriplePattern or BGP, because we have defined the translation method only for those two.
> 
> However, it seems that translating unions (of TriplePatterns and BGPs) is straightforward, right? In which case we could leave that code, and maybe mention it in the thesis, if it is not too long.

I think I can explain in the text how translating a more complex subquery such as a union would be done. Because as long as the triples in the union's subpaterns are handled seperately the translation itself would be the same.

**Comment by scferrada** [comment]
Ok, we can do that then :D!

**Review by hartig — CHANGES_REQUESTED** [review]
Besides the smaller issues mentioned in my comments in the code, there are two additional things that need to be added:

1/ The if-else-if-... part in `performRequest` does not cover all relevant cases. There are also two classes that implement `SPARQLGraphPattern` directly, namely `GenericSPARQLGraphPatternImpl1` and `GenericSPARQLGraphPatternImpl2`. These classes may be used in requests to SPARQL endpoints and, thus, need to be handled here. Both of them have a `asJenaOp()` function that returns the graph pattern in the form of a Jena `Op` object. In particular, there are different relevant subclasses of `Op`, namely:
* `OpJoin` (should be handled like a `SPARQLGroupPattern`),
* `OpUnion` (should be handled like a `SPARQLUnionPattern`),
* `OpFilter` (tricky, I am not sure what to do with it -- let's ignore it for the moment),
* `OpBGP` (should be handled like a `BGP`),
* `OpSequence` (should be handled like a `SPARQLGroupPattern`).

2/ As I had mentioned in your other PR (#177), it is problematic to change `PhysicalOpRequest` as you do here. Instead, the new class `PhysicalOpRequestWithTranslation` that I had added to  PR #177 needs to be used here as well. Maybe, before you start redoing PR #177 (for which we said you need to include the new class `PhysicalOpRequestWithTranslation`), you include this class here and then start the redoing of PR #177 only after this PR here is merged. Note that it is not only that one new class, some other classes where affected too; it is all in this one commit that I had made in PR #177.

File: src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpRequestSPARQLWithTranslation.java [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+{
+	public ExecOpRequestSPARQLWithTranslation( final SPARQLRequest req, final SPARQLEndpoint fm ) {
+		super( req, fm );
+	}
+
+	@Override
+	protected void process( final SolMapsResponse response, final IntermediateResultElementSink sink )
+	{
+		for ( SolutionMapping sm : response.getSolutionMappings() ) {
+			for (SolutionMapping smTranslated : fm.getVocabularyMapping().translateSolutionMapping(sm)) {
```
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpRequestSPARQLWithTranslation.java:48]:
```suggestion
		for ( final SolutionMapping sm : response.getSolutionMappings() ) {
			for ( final SolutionMapping smTranslated : fm.getVocabularyMapping().translateSolutionMapping(sm) ) {
```

File: src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpRequestSPARQLWithTranslation.java [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+		} else if (query instanceof SPARQLGroupPattern) {
+			reqTranslation = handleGroup((SPARQLGroupPattern) query);
+		} else if (query instanceof SPARQLUnionPattern) {
+			reqTranslation = handleUnion((SPARQLUnionPattern) query);
+		} else if (query instanceof GenericSPARQLGraphPatternImpl1) {
+			Op gp = ((GenericSPARQLGraphPatternImpl1) query).asJenaOp();
+			reqTranslation = handleOp(gp);		
+		} else if (query instanceof GenericSPARQLGraphPatternImpl2) {
+			Op gp = ((GenericSPARQLGraphPatternImpl2) query).asJenaOp();
+			reqTranslation = handleOp(gp);
```
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpRequestSPARQLWithTranslation.java:72]:
```suggestion
		final SPARQLGraphPattern reqTranslation;
		if (query instanceof TriplePattern) {
			reqTranslation = handleTriplePattern((TriplePattern) query);	
		} 
		else if (query instanceof BGP) {
			reqTranslation = handleBGP((BGP) query);
		}
		else if (query instanceof SPARQLGroupPattern) {
			reqTranslation = handleGroup((SPARQLGroupPattern) query);
		}
		else if (query instanceof SPARQLUnionPattern) {
			reqTranslation = handleUnion((SPARQLUnionPattern) query);
		}
		else if (query instanceof GenericSPARQLGraphPatternImpl1) {
			final Op gp = ((GenericSPARQLGraphPatternImpl1) query).asJenaOp();
			reqTranslation = handleOp(gp);		
		}
		else if (query instanceof GenericSPARQLGraphPatternImpl2) {
			final Op gp = ((GenericSPARQLGraphPatternImpl2) query).asJenaOp();
			reqTranslation = handleOp(gp);	
```

File: src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpRequestSPARQLWithTranslation.java [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+			throw new FederationAccessException("Unsupported type of pattern: " + query.getClass().getName(), req, fm);
+		}
+		
+		final SPARQLRequest newReq = new SPARQLRequestImpl(reqTranslation);
+		return FederationAccessUtils.performRequest(fedAccessMgr, newReq, fm);	
+	}
+	
+	
+	
+	protected SPARQLGraphPattern handleTriplePattern(TriplePattern tp) {
```
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpRequestSPARQLWithTranslation.java:84]:
```suggestion
	
	
	protected SPARQLGraphPattern handleTriplePattern( final TriplePattern tp ) {
```

File: src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpRequestSPARQLWithTranslation.java [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+		return fm.getVocabularyMapping().translateTriplePattern((TriplePattern) tp);
+	}
+	
+	protected SPARQLGraphPattern handleBGP(BGP bgp) {
```
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpRequestSPARQLWithTranslation.java:88]:
```suggestion
	protected SPARQLGraphPattern handleBGP( final BGP bgp ) {
```

File: src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpRequestSPARQLWithTranslation.java [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+
+		if ( allSubPatternsAreTriplePatterns ) {
+			return new BGPImpl(tpSubPatterns);
+		}
+		else {
+			return new SPARQLGroupPatternImpl(allSubPatterns);
+		}
+	}
+	
+	protected SPARQLGraphPattern handleUnion(SPARQLUnionPattern up) throws FederationAccessException {
```
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpRequestSPARQLWithTranslation.java:113]:
```suggestion
	protected SPARQLGraphPattern handleUnion( final SPARQLUnionPattern up ) throws FederationAccessException {
```

File: src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpRequestSPARQLWithTranslation.java [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+			if (i instanceof TriplePattern) {
+				unionTranslation.addSubPattern(handleTriplePattern((TriplePattern) i));
+			} else if (i instanceof BGP) {
+				unionTranslation.addSubPattern(handleBGP((BGP) i));
+			} else if (i instanceof SPARQLUnionPattern) {
+				unionTranslation.addSubPattern(handleUnion((SPARQLUnionPattern) i));
+			} else if (i instanceof SPARQLGroupPattern) {
+				unionTranslation.addSubPattern(handleGroup((SPARQLGroupPattern) i));
+			} else {
+				throw new FederationAccessException(i.toString(), req, fm);
```
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpRequestSPARQLWithTranslation.java:125]:
```suggestion
				throw new FederationAccessException("Unsupported type of pattern: " + i.getClass().getName(), req, fm);
```

File: src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpRequestSPARQLWithTranslation.java [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+			}
+		}
+		return unionTranslation;
+	}
+	
+	protected SPARQLGraphPattern handleGroup(SPARQLGroupPattern gp) throws FederationAccessException {
```
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpRequestSPARQLWithTranslation.java:131]:
```suggestion
	protected SPARQLGraphPattern handleGroup( final SPARQLGroupPattern gp ) throws FederationAccessException {
```

File: src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpRequestSPARQLWithTranslation.java [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+		List<SPARQLGraphPattern> subPatterns = new ArrayList<>();
```
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpRequestSPARQLWithTranslation.java:132]:
```suggestion
		final List<SPARQLGraphPattern> subPatterns = new ArrayList<>();
```

File: src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpRequestSPARQLWithTranslation.java [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+			if (i instanceof TriplePattern) {
+				subPatterns.add(handleTriplePattern((TriplePattern) i));
+			} else if (i instanceof BGP) {
+				subPatterns.add(handleBGP((BGP) i));
+			} else if (i instanceof SPARQLUnionPattern) {
+				subPatterns.add(handleUnion((SPARQLUnionPattern) i));
+			} else if (i instanceof SPARQLGroupPattern) {
+				subPatterns.add(handleGroup((SPARQLGroupPattern) i));
+			} else {
+				throw new FederationAccessException(i.toString(), req, fm);
```
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpRequestSPARQLWithTranslation.java:143]:
```suggestion
				throw new FederationAccessException("Unsupported type of pattern: " + i.getClass().getName(), req, fm);
```

File: src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpRequestSPARQLWithTranslation.java [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+			}
+		}
+		
+		return new SPARQLGroupPatternImpl(subPatterns);
+	}
+	
+	protected SPARQLGraphPattern handleOp(Op op) throws FederationAccessException {
```
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpRequestSPARQLWithTranslation.java:150]:
```suggestion
	protected SPARQLGraphPattern handleOp( final Op op ) throws FederationAccessException {
```

File: src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpRequestSPARQLWithTranslation.java [hunk 11]
```diff
@@ -0,0 +1,223 @@
...
+			throw new FederationAccessException("Unsupported type of pattern: " + op.getClass().getName(), req, fm);
+		}
+	}
+	
+	protected SPARQLGraphPattern handleJoin(OpJoin oj) throws FederationAccessException {
+		List<SPARQLGraphPattern> subPatterns = new ArrayList<>();
+		Op left = oj.getLeft();
+		subPatterns.add(handleOp(left));
+		Op right = oj.getRight();
+		subPatterns.add(handleOp(right));
```
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpRequestSPARQLWithTranslation.java:171]:
```suggestion
	protected SPARQLGraphPattern handleJoin( final OpJoin oj ) throws FederationAccessException {
		final List<SPARQLGraphPattern> subPatterns = new ArrayList<>();
		final Op left = oj.getLeft();
		subPatterns.add(handleOp(left));
		final Op right = oj.getRight();
		subPatterns.add(handleOp(right));
```

File: src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpRequestSPARQLWithTranslation.java [hunk 12]
```diff
@@ -0,0 +1,223 @@
...
+		return new SPARQLGroupPatternImpl(subPatterns);
+	}
+	
+	protected SPARQLGraphPattern handleOpUnion(OpUnion ou) throws FederationAccessException {
+		final SPARQLUnionPatternImpl unionTranslation = new SPARQLUnionPatternImpl();
+		Op left = ou.getLeft();
+		unionTranslation.addSubPattern(handleOp(left));
+		Op right = ou.getRight();
```
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpRequestSPARQLWithTranslation.java:179]:
```suggestion
	protected SPARQLGraphPattern handleOpUnion( final OpUnion ou ) throws FederationAccessException {
		final SPARQLUnionPatternImpl unionTranslation = new SPARQLUnionPatternImpl();
		final Op left = ou.getLeft();
		unionTranslation.addSubPattern(handleOp(left));
		final Op right = ou.getRight();
```

File: src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpRequestSPARQLWithTranslation.java [hunk 13]
```diff
@@ -0,0 +1,223 @@
...
+		unionTranslation.addSubPattern(handleOp(right));
+		return unionTranslation;
+	}
+	
+	//TODO: Handle OpFilter
+	protected SPARQLGraphPattern handleOpFilter(OpFilter of) throws FederationAccessException {
+		return null;
```
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpRequestSPARQLWithTranslation.java:186]: No!! Simply returning `null` here is a guarantee for giving us headaches later when we all of a sudden have an unexpected `null` somewhere in a rewritten query pattern. To make sure we will always know that there was an unhandled case, throw an exception here (or remove this method and the corresponding case in the `handleOp(Op)` function).

File: src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpRequestSPARQLWithTranslation.java [hunk 14]
```diff
@@ -0,0 +1,223 @@
...
+	}
+	
+	protected SPARQLGraphPattern handleOpBGP(OpBGP obgp) throws FederationAccessException {
```
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpRequestSPARQLWithTranslation.java:189]:
```suggestion
	protected SPARQLGraphPattern handleOpBGP( final OpBGP obgp ) throws FederationAccessException {	
```

File: src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpRequestSPARQLWithTranslation.java [hunk 15]
```diff
@@ -0,0 +1,223 @@
...
+		return null;
+	}
+	
+	protected SPARQLGraphPattern handleOpBGP(OpBGP obgp) throws FederationAccessException {	
+		final List<SPARQLGraphPattern> allSubPatterns = new ArrayList<>();
+		final Set<TriplePattern> tpSubPatterns = new HashSet<>();
+		boolean allSubPatternsAreTriplePatterns = true; // assume yes
+
+		for( final Triple i : obgp.getPattern().getList() ) {
+			TriplePattern itp = new TriplePatternImpl(i);
```
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpRequestSPARQLWithTranslation.java:195]:
```suggestion
			final TriplePattern itp = new TriplePatternImpl(i);
```

File: src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpRequestSPARQLWithTranslation.java [hunk 16]
```diff
@@ -0,0 +1,223 @@
...
+		if ( allSubPatternsAreTriplePatterns ) {
+			return new BGPImpl(tpSubPatterns);
+		}
+		else {
+			return new SPARQLGroupPatternImpl(allSubPatterns);
+		}
+	}
+	
+	protected SPARQLGraphPattern handleOpSequence(OpSequence os) throws FederationAccessException {
+		List<SPARQLGraphPattern> subPatterns = new ArrayList<>();
```
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpRequestSPARQLWithTranslation.java:216]:
```suggestion
	protected SPARQLGraphPattern handleOpSequence( final OpSequence os ) throws FederationAccessException {
		final List<SPARQLGraphPattern> subPatterns = new ArrayList<>();
```

File: src/main/java/se/liu/ida/hefquin/engine/queryplan/physical/impl/PhysicalOpRequest.java [hunk 17]
```diff
@@ -43,6 +44,9 @@ public NullaryExecutableOp createExecOp( final ExpectedVariables ... inputVars )
 		final ReqType req = lop.getRequest();
 		final MemberType fm = lop.getFederationMember();
 		if ( fm instanceof SPARQLEndpoint && req instanceof SPARQLRequest ) {
+			if(fm.getVocabularyMapping() != null) {
+				return new ExecOpRequestSPARQLWithTranslation( (SPARQLRequest) req, (SPARQLEndpoint) fm );
+			}
```
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/queryplan/physical/impl/PhysicalOpRequest.java:49]: No. Please remove that. This case will be covered by the new `PhysicalOpRequestWithTranslation` class.

**Review by hartig — CHANGES_REQUESTED** [review]
Great. All my comments are addressed. Now, I only have a few additional comments; mostly improvements to the code. You can commit all my change suggestions together in a batch (there is a feature for that on the github page for the PR).

**Review by hartig — APPROVED** [review]
Now this PR is ready to merge. Thanks!

**Pull request merged**